sd_mining_core/utils/file_utils.py

```diff
@@ -27,61 +56,141 @@ def download_file(base_dir, file_url, file_name):
+                ) and (not config.exclude_sdxl or not model['type'].startswith('sdxl'))
+            }
+            config.lora_configs = { 
+                lora['name']: lora for lora in loras 
+            }
 
-        config.lora_configs = { 
-            lora['name']: lora
-            for lora in loras 
+        config.vae_configs = {
+            vae['name']: vae for vae in vaes
         }
 
         total_size = 0
         files_to_download = []
-        for model in chain(config.model_configs.values(), config.lora_configs.values()):
-            if 'type' not in model or (model['type'] not in ['sd15', 'sdxl10', 'vae', 'lora']):
-                continue
-            if not 'size_mb' in model:
-                print(f"Warning: Model {model['name']} does not have a size_mb field. models.json is misconfgured. Skipping.")
-                continue
-            file_path = os.path.join(config.base_dir, model['name'] + ".safetensors")
-            if not os.path.exists(file_path):
-                size_mb = model['size_mb']
-                total_size += size_mb
-                files_to_download.append(model)
-            vae_name = model.get('vae', None)
-            if vae_name is not None:
-                vae_path = os.path.join(config.base_dir, model['vae'] + ".safetensors")
+        
+        def process_model(model):
+            nonlocal total_size, files_to_download
+            if 'type' not in model or (model['type'] not in ['sd15', 'sdxl10', 'vae', 'lora', 'flux-dev', 'composite15', 'compositexl']):
+                return
+            
+            if model['type'] in ['composite15', 'compositexl']:
+                # Handle composite models
+                base_model = config.model_configs.get(model['base'])
+                if base_model:
+                    process_model(base_model)  # Process the base model
+                else:
+                    logging.warning(f"Base model '{model['base']}' for composite model '{model['name']}' not found.")
+                
+                # Process the LoRA part
+                lora = config.lora_configs.get(model['name'])
+                if lora:
+                    process_model(lora)
+                else:
+                    logging.warning(f"LoRA part for composite model {model['name']} not found.")
+                return
+
+            if 'size_mb' not in model:
+                logging.warning(f"Model {model['name']} does not have a size_mb field. models.json is misconfigured. Skipping.")
+                return
+            
+            if model['type'] == 'flux-dev':
+                if not check_flux_dev_files(config.base_dir, config.flux_dev_file_downloads):
+                    total_size += model['size_mb']
+                    files_to_download.append(model)
+            else:
+                file_path = os.path.join(config.base_dir, model['name'] + ".safetensors")
+                if not os.path.exists(file_path):
```

Comment:
After the refactor, using `os.path.exists` may result in duplicate models being downloaded. Think this case
Model 1: SDXL Base + Lora 1
Model 2: SDXL Base

then files_to_download list will include duplicate "SDXL base" - is it right?

I think we should check whether files_to_download list includes the target model